src/lepton/jpgcoder.cc

```diff
@@ -1583,6 +1583,45 @@ struct caching_ibytestream
 std::vector<unsigned char> caching_ibytestream::g_input_data;
 
 
+namespace Sirikata {
+// ZBR: memory-caching implementation of Sirikata::DecoderReader API
+class CachingDecoderReader : public DecoderReader
+{
+    static std::vector<unsigned char> g_input_data;  // buffer holding entire contents of the current input file. We made it global in order to avoid memory reallocation for each next file
+    size_t input_pos;  // current reading position in the input data
+    
+public:
+    CachingDecoderReader (DecoderReader* base_stream)
+    {
+        input_pos = 0;
+        
+        // Read the entire input file into g_input_data[]
+        const size_t CHUNK_SIZE = 1024*1024;  // 1 MiB looks like a reasonable read chunk
+        size_t cursize = 0;
+        for(;;) {
+            g_input_data.resize (cursize + CHUNK_SIZE);
+            auto bytes = IOUtil::ReadFull (base_stream, g_input_data.data()+cursize, CHUNK_SIZE);
```

Comment:
Why isn't `base_stream->Read()` used here?

Reply:
It's employed internally by ReadFull: https://github.com/dropbox/lepton/blob/7533355cfc83bcb7d5a3e4947453842d41199c8b/src/io/ioutil.hh#L22

I done that as quick&dirty hack allowing to avoid dealing with read errors and repeating reads until end of file (which may be required for networking I/O afaik).